[Parse] Allow prefix operators containing `/` with `/.../` regex literals

hamishknight opened the pull request:
When lexing a regex literal, treat a prefix operator containing `/` the same as the unapplied infix operator case, where we tentatively lex. This means that we bail if there is no closing `/` or the starting character is invalid. This leaves binary operator containing `/` in expression position as the last place where we know that we definitely have a regex literal.

Additionally, expand the starting `)` character heuristic to cover the entire range of the regex literal, ensuring to take escapes and custom character classes into account. This means we will not lex a regex literal if there is an unbalanced `)` somewhere in the body.

hamishknight commented:
@swift-ci please test

hamishknight commented:
Source compat running on https://github.com/apple/swift/pull/42250

rintaro commented on lib/Parse/ParseExpr.cpp:
```diff
@@ -3226,7 +3234,7 @@ ParserStatus Parser::parseExprList(tok leftTok, tok rightTok,
     // need to do this before handling unapplied operator references, as e.g
     // `(/, /)` might be a regex literal.
-    tryLexRegexLiteral(/*mustBeRegex*/ false);
+    tryLexRegexLiteral(/*forUnappliedOperator*/ true);
```
`forUnappliedOperator` might be confusing. More like `canBeUnappliedOperator` because we don't even know there's a `/` in this token at this point.

hamishknight replied:
Actually changed it so we don't call `tryLexRegexLiteral` unless we know we have an unapplied operator. This lets us avoid re-lexing twice for prefix operators, as they only need to be re-lexed in `parseExprUnary`. How does it look?

rintaro commented on test/StringProcessing/Parse/forward-slash-regex.swift:
```diff
@@ -309,17 +308,84 @@ quxqux(!/^/)
+
+let fn: (Int, Int) -> Int = (/)
+
+_ = /\()/
+// expected-error@-1 {{'/' is not a prefix unary operator}}
+// expected-error@-2 {{'/' is not a postfix unary operator}}
+// expected-error@-3 {{invalid component of Swift key path}}
+
```
Could you add tests for `\)` being ignored? e.g. `_ = (/whatever\)/` should be a regex inside a paren expression with missing `)`

hamishknight replied:
There is a test for `/\)/`, but added your example too, thanks!

hamishknight committed:
[Parse] Re-allow prefix operators containing `/`

hamishknight committed:
[Parse] Lenient prefix `/` parsing

Treat a prefix operator containing `/` the same as
the unapplied infix operator case, where we
tentatively lex. This means that we bail if there
is no closing `/` or the starting character is
invalid. This leaves binary operator containing
`/` in expression position as the last place where
we know that we definitely have a regex literal.

hamishknight committed:
[Parse] Expand unbalanced `)` regex literal heuristic

Previously we would only check for a starting
character of `)` when performing a tentative
lex of a regex literal. Expand this to cover the
entire range of the regex literal, ensuring to
take escapes and custom character classes into
account.

hamishknight commented:
@swift-ci please test

rintaro reviewed (approved):
LGTM
